CDAP-6372 add spark streaming app

albertshau opened the pull request:
Added apis for new StreamingSource and Windower plugin
types for the data streams application.

Also doing some more refactoring on the spark program so that both
the data streams and data pipeline apps can share almost all the
same logic. Introduced the SparkCollection and SparkPairCollection
interfaces which the main driver performs operations on.
The data pipeline app implements them with RDDs while the data
streams app implements it with DStreams.

albertshau committed:
CDAP-6372 add spark streaming app

Added apis for new StreamingSource and Windower plugin
types for the data streams application.

Also doing some more refactoring on the spark program so that both
the data streams and data pipeline apps can share almost all the
same logic. Introduced the SparkCollection and SparkPairCollection
interfaces which the main driver performs operations on.
The data pipeline app implements them with RDDs while the data
streams app implements it with DStreams.

sreevatsanraman commented on cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsApp.java:
```diff
@@ -0,0 +1,45 @@
+    PipelineSpecGenerator<DataStreamsConfig, DataStreamsPipelineSpec> specGenerator =
+      new DataStreamsPipelineSpecGenerator(
+        getConfigurer(),
+        ImmutableSet.of(StreamingSource.PLUGIN_TYPE),
```
Are there no transformations on DataStreams?

sreevatsanraman replied:
Ignore this comment. I see that we have the other types.

sreevatsanraman commented on cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsPipelineSpecGenerator.java:
```diff
@@ -0,0 +1,51 @@
+      batchIntervalMillis = TimeParser.parseDuration(config.getBatchInterval());
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
+        String.format("Unable to parse batchInterval '%s'", config.getBatchInterval()), e);
```
Let's make sure we won't return the exception in the REST response incase of config issues. The message should be sufficient.

sreevatsanraman commented on cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java:
```diff
@@ -0,0 +1,90 @@
+  public void beforeSubmit(SparkClientContext context) throws Exception {
+    SparkConf sparkConf = new SparkConf();
+    sparkConf.set("spark.driver.extraJavaOptions", "-XX:MaxPermSize=256m");
+    sparkConf.set("spark.executor.extraJavaOptions", "-XX:MaxPermSize=256m");
```
We should make these two configurations configurable.

sreevatsanraman commented on cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java:
```diff
@@ -0,0 +1,527 @@
+  }
+
+  @Test
+  public void testWindower() throws Exception {
```
Is Windowed join supported? We should add a testcase if it is.

albertshau replied:
There is no windowed join capability in a single stage. To do it, somebody would first window then join, in which case the two plugin types can be tested separately.

sreevatsanraman commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/batch/RDDCollection.java:
```diff
@@ -0,0 +1,127 @@
+    SparkExecutionPluginContext sparkPluginContext =
+      new BasicSparkExecutionPluginContext(sec, jsc, datasetContext, stageName);
+
+    // TODO: figure out how to do this in a better way...
```
Can we add JIRAs for these.

sreevatsanraman commented:
Have minor comments. Fix and merge. LGTM

albertshau committed:
CDAP-6372 add java opts config option

Making the extra java opts a configuration setting instead of
hardcoding it. Also adding jira references to records in and out.

albertshau commented:
http://builds.cask.co/browse/CDAP-DUT4509
